src/coord/src/catalog.rs

```diff
@@ -1868,6 +1880,57 @@ impl sql::catalog::Catalog for ConnCatalog<'_> {
     fn config(&self) -> &sql::catalog::CatalogConfig {
         &self.catalog.config
     }
+
+    fn minimal_qualification(&self, full_name: &FullName) -> PartialName {
+        let database = match &full_name.database {
+            DatabaseSpecifier::Ambient => None,
+            DatabaseSpecifier::Name(n) => {
+                if *n == self.database {
+                    None
+                } else {
+                    Some(n.clone())
+                }
+            }
+        };
+
+        let schema = if self
+            .search_path
+            .iter()
+            .any(|p| p == &full_name.schema.as_str())
+        {
+            // If `full_name.schema` is findable in the `search_path`, find
+            // first reference to `item`.
+            let mut schema = None;
+            for search_schema in self.search_path {
+                let partial_name = PartialName {
+                    database: database.clone(),
+                    schema: Some(search_schema.to_string()),
+                    item: full_name.item.clone(),
+                };
+                // If first reference to `item` in `search_path` is not
+                // `full_name`, the `PartialName` must be schema qualified.
+                if let Ok(item) = self.resolve_item(&partial_name) {
+                    if item.name() != full_name {
+                        schema = Some(full_name.schema.clone());
+                    }
+                    break;
+                }
+            }
+            schema
+        } else {
+            // If `search_path` does not contain `full_name.schema`, the
+            // `PartialName` must contain it.
+            Some(full_name.schema.clone())
+        };
+
+        let res = PartialName {
+            database,
+            schema,
+            item: full_name.item.clone(),
+        };
+        assert!(self.resolve_item(&res).unwrap().name() == full_name);
+        res
+    }
```

Comment:
As I'm wont to do, I tried out all these suggestions to make sure they were legit. Here's a patch!

```diff
diff --git a/src/coord/src/catalog.rs b/src/coord/src/catalog.rs
index b268355be..08e3a7ff6 100644
--- a/src/coord/src/catalog.rs
+++ b/src/coord/src/catalog.rs
@@ -26,7 +26,7 @@ use expr::{GlobalId, IdHumanizer, OptimizedRelationExpr, ScalarExpr};
 use repr::{RelationDesc, ScalarType};
 use sql::ast::display::AstDisplay;
 use sql::ast::Expr;
-use sql::catalog::CatalogError as SqlCatalogError;
+use sql::catalog::{Catalog as SqlCatalog, CatalogError as SqlCatalogError};
 use sql::names::{DatabaseSpecifier, FullName, PartialName, SchemaName};
 use sql::plan::{Params, Plan, PlanContext};
 use transform::Optimizer;
@@ -79,12 +79,12 @@ pub const FIRST_USER_OID: u32 = 20_000;
 pub struct Catalog {
     by_name: BTreeMap<String, Database>,
     by_id: BTreeMap<GlobalId, CatalogEntry>,
+    by_oid: HashMap<u32, GlobalId>,
     indexes: HashMap<GlobalId, Vec<(GlobalId, Vec<ScalarExpr>)>>,
     ambient_schemas: BTreeMap<String, Schema>,
     temporary_schemas: HashMap<u32, Schema>,
     storage: Arc<Mutex<storage::Connection>>,
     oid_counter: u32,
-    oid_to_id: HashMap<u32, GlobalId>,
     config: sql::catalog::CatalogConfig,
 }
 
@@ -396,12 +396,12 @@ impl Catalog {
         let mut catalog = Catalog {
             by_name: BTreeMap::new(),
             by_id: BTreeMap::new(),
+            by_oid: HashMap::new(),
             indexes: HashMap::new(),
             ambient_schemas: BTreeMap::new(),
             temporary_schemas: HashMap::new(),
             storage: Arc::new(Mutex::new(storage)),
             oid_counter: FIRST_USER_OID,
-            oid_to_id: HashMap::new(),
             config: sql::catalog::CatalogConfig {
                 startup_time: SystemTime::now(),
                 nonce: rand::random(),
@@ -794,10 +794,6 @@ impl Catalog {
         &self.by_id[id]
     }
 
-    pub fn oid_to_id(&self, oid: &u32) -> &GlobalId {
-        &self.oid_to_id[oid]
-    }
-
     /// Creates a new schema in the `Catalog` for temporary items
     /// indicated by the TEMPORARY or TEMP keywords.
     pub fn create_temporary_schema(&mut self, conn_id: u32) -> Result<(), Error> {
@@ -927,7 +923,7 @@ impl Catalog {
             .expect("catalog out of sync");
         let schema_id = schema.id;
         schema.items.insert(entry.name.item.clone(), entry.id);
-        self.oid_to_id.insert(oid, entry.id.clone());
+        self.by_oid.insert(oid, entry.id);
         self.by_id.insert(entry.id, entry);
 
         Event::CreatedItem {
@@ -1757,7 +1753,13 @@ impl From<PlanContext> for SerializedPlanContext {
     }
 }
 
-impl sql::catalog::Catalog for ConnCatalog<'_> {
+impl ConnCatalog<'_> {
+    fn resolve_item_name(&self, name: &PartialName) -> Result<&FullName, SqlCatalogError> {
+        self.resolve_item(name).map(|entry| entry.name())
+    }
+}
+
+impl SqlCatalog for ConnCatalog<'_> {
     fn search_path(&self, include_system_schemas: bool) -> Vec<&str> {
         if include_system_schemas {
             self.search_path.to_vec()
@@ -1828,8 +1830,8 @@ impl sql::catalog::Catalog for ConnCatalog<'_> {
     }
 
     fn get_item_by_oid(&self, oid: &u32) -> &dyn sql::catalog::CatalogItem {
-        let id = self.catalog.oid_to_id(oid);
-        self.catalog.get_by_id(id)
+        let id = self.catalog.by_oid[oid];
+        self.catalog.get_by_id(&id)
     }
 
     fn item_exists(&self, name: &FullName) -> bool {
@@ -1893,33 +1895,15 @@ impl sql::catalog::Catalog for ConnCatalog<'_> {
             }
         };
 
-        let schema = if self
-            .search_path
-            .iter()
-            .any(|p| p == &full_name.schema.as_str())
+        let schema = if database.is_none()
+            && self.resolve_item_name(&PartialName {
+                database: None,
+                schema: None,
+                item: full_name.item.clone(),
+            }) == Ok(full_name)
         {
-            // If `full_name.schema` is findable in the `search_path`, find
-            // first reference to `item`.
-            let mut schema = None;
-            for search_schema in self.search_path {
-                let partial_name = PartialName {
-                    database: database.clone(),
-                    schema: Some(search_schema.to_string()),
-                    item: full_name.item.clone(),
-                };
-                // If first reference to `item` in `search_path` is not
-                // `full_name`, the `PartialName` must be schema qualified.
-                if let Ok(item) = self.resolve_item(&partial_name) {
-                    if item.name() != full_name {
-                        schema = Some(full_name.schema.clone());
-                    }
-                    break;
-                }
-            }
-            schema
+            None
         } else {
-            // If `search_path` does not contain `full_name.schema`, the
-            // `PartialName` must contain it.
             Some(full_name.schema.clone())
         };
 
@@ -1928,7 +1912,7 @@ impl sql::catalog::Catalog for ConnCatalog<'_> {
             schema,
             item: full_name.item.clone(),
         };
-        assert!(self.resolve_item(&res).unwrap().name() == full_name);
+        assert_eq!(self.resolve_item_name(&res), Ok(full_name));
         res
     }
 }
diff --git a/src/sql/src/catalog.rs b/src/sql/src/catalog.rs
index aa3ede0ea..81b0d3cf7 100644
--- a/src/sql/src/catalog.rs
+++ b/src/sql/src/catalog.rs
@@ -243,7 +243,7 @@ impl fmt::Display for CatalogItemType {
 }
 
 /// An error returned by the catalog.
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 pub enum CatalogError {
     /// Unknown database.
     UnknownDatabase(String),
```